feat(draco): Support FLOAT64, INT64, UINT64 attribute types

ibgreen opened the pull request:
For #2947 and https://github.com/visgl/deck.gl/issues/8746
- I can't find support for these types in the Draco WASM encoder interfaces.
- I don't have a test file, and due to lack of encoding support I can't even write roundtrip encode/decode tests.

@donmccurdy

donmccurdy commented:
The Draco extension for glTF that I'm familiar with (`KHR_draco_mesh_compression`) does not allow 64-bit or boolean types, so I'm not sure how to produce a draco-compressed file to test this either.

It looks like the data in #8746 is a 3D Tiles v1.0 tileset, using .pnts and `3DTILES_draco_point_compression`. The `3DTILES_draco_point_compression` specification is deprecated with a [migration guide](https://github.com/CesiumGS/3d-tiles/blob/main/specification/TileFormats/glTF/MIGRATION.adoc), and I believe for compressed point clouds the recommended workaround would be to use Meshopt compression. I'm unsure whether Meshopt supports 64-bit precision either.

/cc @javagl might know better how we should proceed here, or how we could obtain a simplified (.glb?) file to test 64-bit decompression.

javagl commented:
I might be lacking some context (and am not entirely sure what the question is).

On the _creating_ side, encoding `double` (aka `DT_FLOAT64` values) should be possible with [`PointCloudBuilder::addAttribute`](https://github.com/google/draco/blob/7d58126d076bc3f5f9d8c114d1700b7311faecfe/src/draco/point_cloud/point_cloud_builder.cc#L29), passing in `DT_FLOAT64` as the last parameter.

On the _consuming_ side, there seem to be the [`GetAttribute...ForAllPoints`](https://github.com/google/draco/blob/7d58126d076bc3f5f9d8c114d1700b7311faecfe/src/draco/javascript/emscripten/decoder_webidl_wrapper.h#L179) functions. From some snippets, it _looks_ like `GetAttributeFloatForAllPoints` might convert `double` to `float` under the hood. There also seems to be a generic `GetAttributeDataArrayForAllPoints` for arbitrary data types, but I'd have to look up how this could be used in JS.

This refers to PNTS with the `3DTILES_draco_point_compression` extension, though. I don't fully understand where glTF/GLB comes into play here. When the properties are stored as glTF vertex attributes, they cannot be 64 bit values anyhow. And when they are 32bit, then it _should_ be possible to generically apply Meshopt (but I'd have to look up many further details here as well...).

If this is not the point, then a bit more context could be helpful.

(An aside: A file that contains PNTS data with a 64bit Draco-encoded property is part of the tileset that was linked in https://github.com/visgl/deck.gl/issues/8746, namely the file at https://storage.googleapis.com/external-geo-projects/GTE464_Crocodile_Lower_LAS/2/1.pnts )

donmccurdy commented:
Thanks @javagl!

We've got bug reports from users that they cannot load 3D tileset containing PNTS data, which appear to contain 64-bit accessors compresesd with Draco, produced by Cesium Ion.

I think I'm hoping to learn:

1. Do you have any opinion on whether the wider community _should_ support PNTS at this point, or is it on a deprecation path with Cesium Ion? Should we be pointing users to a migration path, and if so what is that migration path? We'd prefer not to waste effort, which is in limited supply, perhaps we should be spending time supporting the new thing rather than the old?
2. If supporting PNTS is safe enough for the foreseeable future, are there public sample files containing 64-bit Draco-compressed data that we could use for testing?

ibgreen commented:
Yes, it is a bit inconvenient to work with that PNTS file for Draco Loader unit testing, as it is a full 3D tiles file. We do have the code in loaders.gl to extract it, but it increases the effort. 

Another strategy could be to just ignore this type of attributes, instead of crashing - if none of the required geometry attributes can be 64 bits, this might just be an extra attribute.

ibgreen commented:
loaders.gl has had solid support for cesium point clouds for ~5 years, in contrast to the mesh based tiles, there has been very little maintenance required.

See https://loaders.gl/examples/3d-tiles.  

So this issue came a bit out of the blue...

javagl commented:
> Do you have any opinion on whether the wider community should be adding PNTS in 2024, or is it on a deprecation path with Cesium Ion?

The tile formats like PNTS are 'deprecated', in favor of glTF. There are different ways of interpreting that, and different conclusions that implementors might draw from that. 

I usually try to avoid 'opinions', but ... _somewhat_ subjectively: If someone now wanted to start implementing an application that could convert LAS (point cloud) files into 3D Tiles, then I'd suggest to not generate PNTS any more, but generate glTF/GLB instead. The main reasons, roughly: Better interoperability, more tooling, and easier support for rendering that on the _client_ side. There is no really compelling reason to still _generate_ PNTS.

Now, this issue here _does_ refer to the _client_ side. And there, I'd hesitate to give a similar recommendation. Cesium ion is still creating PNTS data for point clouds. (And I'm not giving away a secret here: Throw some LAS into ion, load the result, look at the network tab - there's the `.pnts`). The format won't go away too soon. And people will probably expect clients to support 3D Tiles with PNTS when they claim that they "support 3D Tiles". 

All this with the usual disclaimers: I don't know how much effort has to be invested here in order to "fix the issue". But there are probably different degrees:

- Zero effort: Don't support them. Let it crash.
- Little effort: Ignore the 64bit attributes in the reader and writer
- Medium effort: Add basic support in the reader (maybe converting to `float`), but not in the writer
- Higher effort: Fully support reading and writing them as true 64 bit values

I'm making some guesses here, based on the "Draco Attribute Support / `DracoLoader` | `DracoWriter`" table in the changed files of this PR, and comments like this one:

> Another strategy could be to just ignore this type of attributes, instead of crashing - if none of the required geometry attributes can be 64 bits, this might just be an extra attribute.

That _could_  be a viable option, if these `DT_FLOAT64` attributes cannot "trivially" be supported by some magic conversion that is done with `GetAttributeFloatForAllPoints`. From quickly skimming over the [relevant code in CesiumJS](https://github.com/CesiumGS/cesium/blob/4bd2acb09cdb2bc42083e810138db77ce8c97a46/packages/engine/Source/Workers/decodeDraco.js#L146), it looks like these 64bit are just shrugged off and written into a `Float32Array`. Maybe that could be done here as well....?

> loaders.gl has had solid support for cesium point clouds for ~5 years, in contrast to the mesh based tiles, there has been very little maintenance required.
> ...
> So this issue came a bit out of the blue...

Let's consult my crystal ball here: 🔮 (It's blue, indeed). Recently, Cesium added extended support for metadata from point clouds - specifically, the additional data that is contained in LAS files: https://cesium.com/blog/2024/03/20/preserving-more-metadata-for-point-clouds-using-3dtiles/ . And ... it's exactly that `GpsTime` in the `1.pnts` that I linked to that uses the `DOUBLE` data type and that is compressed with `3DTILES_draco_point_compression` - here's the batch table JSON:
```json
{
  "Classification" : {
    "byteOffset" : 0,
    "componentType" : "UNSIGNED_BYTE",
    "type" : "SCALAR"
  },
  "GpsTime" : {
    "byteOffset" : 0,
    "componentType" : "DOUBLE",
    "type" : "SCALAR"
  },
  "Intensity" : {
    "byteOffset" : 0,
    "componentType" : "UNSIGNED_SHORT",
    "type" : "SCALAR"
  },
  "NumberOfReturns" : {
    "byteOffset" : 0,
    "componentType" : "UNSIGNED_BYTE",
    "type" : "SCALAR"
  },
  "PointSourceID" : {
    "byteOffset" : 0,
    "componentType" : "UNSIGNED_SHORT",
    "type" : "SCALAR"
  },
  "ReturnNumber" : {
    "byteOffset" : 0,
    "componentType" : "UNSIGNED_BYTE",
    "type" : "SCALAR"
  },
  "ScanAngleRank" : {
    "byteOffset" : 0,
    "componentType" : "SHORT",
    "type" : "SCALAR"
  },
  "extensions" : {
    "3DTILES_draco_point_compression" : {
      "properties" : {
        "Classification" : 8,
        "GpsTime" : 7,
        "Intensity" : 2,
        "NumberOfReturns" : 4,
        "PointSourceID" : 6,
        "ReturnNumber" : 3,
        "ScanAngleRank" : 5
      }
    }
  }
}
```
So... that's likely why this came up now...

ibgreen commented:
@javagl thanks for all the context. The LAS export triggering this makes perfect sense, added a link to the blog post in the release notes for this fix. 

I added a test case for the PNTS file you indicated (hopefully the license is OK) and we extract the 64 bit attribute just fine. 

https://github.com/visgl/loaders.gl/blob/ib/draco-attributes/modules/3d-tiles/test/lib/parsers/point-cloud-3d-tile.spec.ts#L127

javagl commented:
> I added a test case for the PNTS file you indicated (hopefully the license is OK)

The link to that PNTS file was just taken from the tileset that was posted in the issue https://github.com/visgl/deck.gl/issues/8746

@jeafreezy Maybe you want to confirm that it is OK to use a _single_ PNTS file from this tileset (like https://storage.googleapis.com/external-geo-projects/GTE464_Crocodile_Lower_LAS/2/1.pnts ) for the unit test here.

If not, I can throw some random PD/CC0 LAS file into Cesium ion and generate such a PNTS file for the test.

Pessimistress commented on modules/draco/src/lib/draco-parser.ts:
```diff
@@ -351,6 +355,9 @@ export default class DracoParser {
   ): {value: TypedArray; size: number} {
     const TypedArrayCtor = DRACO_DATA_TYPE_TO_TYPED_ARRAY_MAP[attribute.data_type];
+    if (!TypedArrayCtor) {
+      throw new Error(`DRACO: Unsupported attribute data type ${attribute.data_type}.`);
```
Seems that any unknown type would belong to some extra property that does not block rendering. Can we warn and ignore instead of throwing?

jeafreezy commented:
> > I added a test case for the PNTS file you indicated (hopefully the license is OK)
> 
> The link to that PNTS file was just taken from the tileset that was posted in the issue [visgl/deck.gl#8746](https://github.com/visgl/deck.gl/issues/8746)
> 
> @jeafreezy Maybe you want to confirm that it is OK to use a _single_ PNTS file from this tileset (like https://storage.googleapis.com/external-geo-projects/GTE464_Crocodile_Lower_LAS/2/1.pnts ) for the unit test here.
> 
> If not, I can throw some random PD/CC0 LAS file into Cesium ion and generate such a PNTS file for the test.

Unfortunately it's not 😞 , kindly proceed to generating a PNTS file for the test.

I will also be revoking the URL, now that the issue has been identified. 

Thank you

ibgreen commented:
> Unfortunately it's not 😞 , kindly proceed to generating a PNTS file for the test.

OK no worries. I will hold off merging the fix until someone contributes a test PNTS to loaders.gl.

ibgreen commented:
Replaced by https://github.com/visgl/loaders.gl/pull/2951

javagl commented:
I'm working on trying to create test data. But you cannot believe the avalance of garbage that is set off by websearching for things like "public domain point cloud las" (etc). In doubt, I'll try to create a dummy LAS file that can be converted into a PNTS with the required Draco/`DOUBLE` attributes.

jeafreezy commented:
@javagl  Yes, I managed to finally receive a go ahead. Please proceed to use it without attribution/licensing issues. Thank you.

ibgreen commented:
@jeafreezy Thanks I am opening a new PR to land this. 

Any chance you could provide the LAS file you used to create this as well? I am unsure if the values I am reading are correct, they look too noisy.

![image](https://github.com/visgl/loaders.gl/assets/7025232/8bfc0f3d-8185-4aeb-9691-c1e8d7d284b4)

javagl commented:
The values should be somewhere around 378550000. 

![Draco double gps time](https://github.com/visgl/loaders.gl/assets/5597569/68184c90-6799-4ca4-88c5-0f745b825a45)

The values that you're seeing ... well, this could be something "simple", like the bytes of  `float` values being interpreted as `double` or vice versa. I see that you're using a `Float64Array` there. As I mentioned earlier: CesiumJS seems to put the values into a `Float32Array`, but ... I have **no** clue about how Draco works, so that's just a (possible meaningless) observation for now.

ibgreen commented:
Looks like you are right, the data is written as Float32. But bizarrely still flagged as FLOAT64: If I parse the attribute that is typed as 64 bit floats as 32 bits then I get this:
![image](https://github.com/visgl/loaders.gl/assets/7025232/34e1f602-4e73-4111-8bd9-288f7a971452)

@shehzan10 Implementing support for your latest PNTS files with 64 bit attributes... Are you able to shed light on why the [draco type isn't just written as DT_FLOAT32](https://github.com/CesiumGS/cesium/blob/4bd2acb09cdb2bc42083e810138db77ce8c97a46/packages/engine/Source/Workers/decodeDraco.js#L146) if the data is converted to and stored as Float32Array?

javagl commented:
I think @lilleyse mentioned that back in the days when all this was introduced, some decisions _might_ have been influenced by lack of 64bit support (in general) in Draco. Some of the questions here might start at the point where this data is _written_ - very roughly: It might be that the implementation that _writes_ the Draco-encoded data chooses to declare it as 64bit, but only writes it out as 32bit. This might then be a bug in the code that writes it, or a "legacy" aspect due to previous limitations of Draco. However, if some of these caveats can be identified and articularted clearly, then they should either be fixed, or some clarifications might be added as 'Implementation Note', in the specification at https://github.com/CesiumGS/3d-tiles/tree/main/extensions/3DTILES_draco_point_compression
